{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import norm\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import mannwhitneyu\n",
    "import seaborn as sns\n",
    "from scipy.optimize import minimize_scalar\n",
    "import argparse\n",
    "import os\n",
    "import pickle\n",
    "from subprocess import call\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR\"\n",
    "os.chdir(work_dir)\n",
    "gene_bed_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/stqtl/gene_merge.bed\"\n",
    "apa_bed_file = \"/mnt/hpc/home/xuxinran/DirectSeq/nano3aqtl/hg19.apadb_v2_final.sorted.bed\"\n",
    "isoform_bed_file = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/irqtl/nano_merge.annotated_transcripts.bed\"\n",
    "pu_bed_File = \"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/puqtl/promoter_final.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bf_to_pvalue(bf, f, trait_type, N, s=None):\n",
    "    \"\"\"\n",
    "    Calculate p-value from Bayes Factor for a SNP given other parameters.\n",
    "\n",
    "    Parameters:\n",
    "    - bf: Bayes Factor (not log)\n",
    "    - f: minor allele frequency (MAF)\n",
    "    - trait_type: 'quant' for quantitative trait or 'cc' for case-control\n",
    "    - N: sample size\n",
    "    - s: proportion of samples that are cases (required if trait_type is 'cc')\n",
    "\n",
    "    Returns:\n",
    "    - Estimated p-value\n",
    "    \"\"\"\n",
    "    def Var_data(f, N):\n",
    "        \"\"\"Variance calculation for quantitative trait.\"\"\"\n",
    "        return 1 / (2 * N * f * (1 - f))\n",
    "    def Var_data_cc(f, N, s):\n",
    "        \"\"\"Variance calculation for case-control data.\"\"\"\n",
    "        return 1 / (2 * N * f * (1 - f) * s * (1 - s))\n",
    "    if trait_type == \"quant\":\n",
    "        sd_prior = 0.15\n",
    "        V = Var_data(f, N)\n",
    "    else:\n",
    "        sd_prior = 0.2\n",
    "        V = Var_data_cc(f, N, s)\n",
    "    r = sd_prior**2 / (sd_prior**2 + V)\n",
    "    def objective(z):\n",
    "        lABF = 0.5 * (np.log(1 - r) + (r * z**2))\n",
    "        return abs(np.exp(lABF) - bf)\n",
    "    # Find the z-score that minimizes the difference between calculated BF and given BF\n",
    "    result = minimize_scalar(objective)\n",
    "    z = abs(result.x)  # Take absolute value as z-score is always positive\n",
    "    # Calculate p-value from z-score\n",
    "    p = 2 * (1 - norm.cdf(z))\n",
    "    return p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSE(A1_m6A, A1_A, A2_m6A, A2_A):\n",
    "    alpha1_post = 1 + A1_m6A\n",
    "    beta1_post = 1 + A1_A\n",
    "    alpha2_post = 1 + A2_m6A\n",
    "    beta2_post = 1 + A2_A\n",
    "    var_A1 = alpha1_post * beta1_post / ((alpha1_post + beta1_post)^2 * (alpha1_post + beta1_post + 1))\n",
    "    var_A2 = alpha2_post * beta2_post / ((alpha2_post + beta2_post)^2 * (alpha2_post + beta2_post + 1))\n",
    "    SE = np.sqrt(var_A1 + var_A2)\n",
    "    return SE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "qtl_dict = {\n",
    "        'inosine-QTL':'/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/Iqtl/nano_merge_I_summary.csv',\n",
    "        'puQTL':'/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/puqtl/nano_merge_promoter_summary.csv',\n",
    "        'm6A-QTL':'/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/m6Aqtl/nano_merge_m6A_summary.csv',\n",
    "        'pseU-QTL':'/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/pseUqtl/nano_merge_pseU_summary.csv',\n",
    "        'm5C-QTL':'/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/m5Cqtl/nano_merge_m5C_summary.csv',\n",
    "        # 'stQTL':'/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/stqtl/nano_merge_stability_summary.csv',\n",
    "        '3aQTL':'/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/3aqtl/nano_merge_APA_summary.csv',\n",
    "        'irQTL':'/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/irqtl/nano_merge_isoform_summary.csv'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inosine-QTL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21943/3887116761.py:30: RuntimeWarning: overflow encountered in exp\n",
      "  return abs(np.exp(lABF) - bf)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "puQTL\n",
      "m6A-QTL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21943/3887116761.py:30: RuntimeWarning: overflow encountered in exp\n",
      "  return abs(np.exp(lABF) - bf)\n",
      "/mnt/hpc/home/xuxinran/mambaforge/envs/methy1/lib/python3.12/site-packages/scipy/optimize/_optimize.py:2502: RuntimeWarning: invalid value encountered in scalar multiply\n",
      "  tmp2 = (x - v) * (fx - fw)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pseU-QTL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_21943/1969462159.py:9: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  qtl_data = pd.read_csv(qtl_file)\n",
      "/tmp/ipykernel_21943/3887116761.py:30: RuntimeWarning: overflow encountered in exp\n",
      "  return abs(np.exp(lABF) - bf)\n",
      "/mnt/hpc/home/xuxinran/mambaforge/envs/methy1/lib/python3.12/site-packages/scipy/optimize/_optimize.py:2502: RuntimeWarning: invalid value encountered in scalar multiply\n",
      "  tmp2 = (x - v) * (fx - fw)\n"
     ]
    }
   ],
   "source": [
    "for qtl_type,qtl_file in qtl_dict.items():\n",
    "    print(qtl_type)\n",
    "    load_and_filter_qtl(qtl_file,qtl_type,work_dir, gene_bed_file, apa_bed_file, isoform_bed_file, pu_bed_File)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# stQTL求SE\n",
    "def analyze_snp_stability_MannWhitneyU(A1_stability, A2_stability):\n",
    "    total_count = len(A1_stability) + len(A2_stability)\n",
    "    A1_count = len(A1_stability)\n",
    "    A2_count = len(A2_stability)\n",
    "    if (total_count < 10) or (A1_count == 0) or (A2_count == 0):\n",
    "        return None,None\n",
    "    stat, p_value = mannwhitneyu(A1_stability, A2_stability, alternative='two-sided')\n",
    "    dominance = 0\n",
    "    for x in A1_stability:\n",
    "        for y in A2_stability:\n",
    "            if x > y:\n",
    "                dominance += 1\n",
    "            elif x < y:\n",
    "                dominance -= 1\n",
    "    delta = dominance / (A1_count * A2_count)\n",
    "    return p_value, delta\n",
    "\n",
    "def bootstrap_delta_se(A1_stability, A2_stability, n_boot=100):\n",
    "    A1_stability = eval(A1_stability)\n",
    "    A2_stability = eval(A2_stability)\n",
    "    delta_values = []\n",
    "    for _ in range(n_boot):\n",
    "        # 从原始数据中随机抽样\n",
    "        A1_sample = np.random.choice(A1_stability, size=len(A1_stability), replace=True)\n",
    "        A2_sample = np.random.choice(A2_stability, size=len(A2_stability), replace=True)\n",
    "        # 计算 delta 值\n",
    "        _, delta = analyze_snp_stability_MannWhitneyU(A1_sample, A2_sample)\n",
    "        delta_values.append(delta)\n",
    "    # 计算 SE\n",
    "    se = np.std(delta_values)\n",
    "    return se"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_and_filter_qtl(qtl_file,qtl_type,work_dir, gene_bed_file, apa_bed_file, isoform_bed_file, pu_bed_File):\n",
    "    if qtl_type == \"stQTL\":\n",
    "        qtl_data = pd.read_csv(qtl_file)\n",
    "        qtl_data['se'] = qtl_data.apply(lambda row: bootstrap_delta_se(row['A1_STscore_l'], row['A2_STscore_l']), axis=1)\n",
    "        new_names = {'pvalue': f'{qtl_type}_pvalue', 'se': f'{qtl_type}_se', 'Beta': f'{qtl_type}_Beta'}\n",
    "        qtl_data = qtl_data.rename(columns=new_names)\n",
    "        qtl_data = qtl_data[[\"chrom\",\"strand\",\"snp_pos\",\"rsID\",\"A1\",\"A2\",\"EAF\",f\"{qtl_type}_Beta\",f\"{qtl_type}_pvalue\",f\"{qtl_type}_se\"]]\n",
    "    elif qtl_type in ['m6A-QTL','pseU-QTL','m5C-QTL','inosine-QTL']:\n",
    "        qtl_data = pd.read_csv(qtl_file)\n",
    "        qtl_data['Pvalue'] = qtl_data.apply(lambda row: bf_to_pvalue(row['BayesFactor'], row['EAF'], \"quant\", 104), axis=1)\n",
    "        qtl_data['se'] = qtl_data.apply(lambda row: getSE(row.iloc[10], row.iloc[8], row.iloc[11], row.iloc[9]), axis=1)\n",
    "        qtl_data = qtl_data.iloc[:, [0, 1, 2, 3, 4, 5, 6, 7, 17, 18, 19]]\n",
    "        qtl_data.columns = [\"chrom\",\"strand\",\"snp_pos\",f\"{qtl_type}_mod_pos\",\"rsID\",\"A1\",\"A2\",\"EAF\",f\"{qtl_type}_Beta\",f\"{qtl_type}_pvalue\",f\"{qtl_type}_se\"]\n",
    "    else:\n",
    "        qtl_data = pd.read_csv(qtl_file)\n",
    "        qtl_data['Pvalue'] = qtl_data.apply(lambda row: bf_to_pvalue(row['BayesFactor'], row['EAF'], \"quant\", 104), axis=1)\n",
    "        qtl_data['se'] = qtl_data.apply(lambda row: getSE(row.iloc[7], row.iloc[9], row.iloc[8], row.iloc[10]), axis=1)\n",
    "        qtl_data = qtl_data.iloc[:, [0,1,2,3,4,5,6,11,15,16,17]]\n",
    "        qtl_data.columns = [\"chrom\",\"strand\",\"snp_pos\",\"rsID\",\"A1\",\"A2\",\"EAF\",f\"{qtl_type}_id\",f\"{qtl_type}_Beta\",f\"{qtl_type}_pvalue\",f\"{qtl_type}_se\"]\n",
    "    qtl_data_copy = qtl_data.copy()\n",
    "    save_esi(qtl_data_copy, qtl_type)\n",
    "    qtl_data_copy = qtl_data.copy()\n",
    "    save_esd(qtl_data_copy, qtl_type)\n",
    "    qtl_data_copy = qtl_data.copy()\n",
    "    save_epi_flist(qtl_data_copy, qtl_type, work_dir, gene_bed_file, apa_bed_file, isoform_bed_file, pu_bed_File)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_esi(df, qtl_type):\n",
    "    df['chrom'] = df['chrom'].str.replace('chr', '', regex=False)\n",
    "    df['rsID'] = df.apply(lambda row: f\"{row['chrom']}:{str(row['snp_pos'])}\" if row['rsID'] == \".\" else row['rsID'], axis=1)\n",
    "    df['LD_pos'] = 0\n",
    "    res_df = df[[\"chrom\",\"rsID\",\"LD_pos\",\"snp_pos\",\"A1\",\"A2\"]]\n",
    "    res_df.to_csv(f'{qtl_type}.esi', index=False, sep=\"\\t\", header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_esd(df, qtl_type):\n",
    "    df['chrom'] = df['chrom'].str.replace('chr', '', regex=False)\n",
    "    df['rsID'] = df.apply(lambda row: f\"{row['chrom']}:{str(row['snp_pos'])}\" if row['rsID'] == \".\" else row['rsID'], axis=1)\n",
    "    if qtl_type in ['m6A-QTL','pseU-QTL','m5C-QTL','inosine-QTL']:\n",
    "        dfs = df.groupby(f\"{qtl_type}_mod_pos\")\n",
    "        for name, group in dfs:\n",
    "            res_df = group[[\"chrom\",\"rsID\",\"snp_pos\",\"A1\",\"A2\",\"EAF\",f\"{qtl_type}_Beta\",f\"{qtl_type}_se\",f\"{qtl_type}_pvalue\"]]\n",
    "            res_df.columns = [\"Chr\",\"SNP\",\"Bp\",\"A1\",\"A2\",\"Freq\",\"Beta\",\"se\",\"p\"]\n",
    "            res_df.to_csv(f'{qtl_type}_{name}.esd', index=False, sep=\"\\t\", header=False)\n",
    "    if qtl_type in ['puQTL','3aQTL','irQTL']:\n",
    "        dfs = df.groupby(f\"{qtl_type}_id\")\n",
    "        for name, group in dfs:\n",
    "            res_df = group[[\"chrom\",\"rsID\",\"snp_pos\",\"A1\",\"A2\",\"EAF\",f\"{qtl_type}_Beta\",f\"{qtl_type}_se\",f\"{qtl_type}_pvalue\"]]\n",
    "            res_df.columns = [\"Chr\",\"SNP\",\"Bp\",\"A1\",\"A2\",\"Freq\",\"Beta\",\"se\",\"p\"]\n",
    "            res_df.to_csv(f'{qtl_type}_{name}.esd', index=False, sep=\"\\t\", header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_epi_flist(df, qtl_type, work_dir,gene_bed_file,apa_bed_file,isoform_bed_file,pu_bed_file):\n",
    "    if qtl_type in ['m6A-QTL','pseU-QTL','m5C-QTL','inosine-QTL']:\n",
    "        df['s'] = df[f\"{qtl_type}_mod_pos\"].astype(int)-1\n",
    "        df['e'] = df[f\"{qtl_type}_mod_pos\"].astype(int)+1\n",
    "        df['score'] = 0\n",
    "        bed_df = df[[\"chrom\",\"s\",\"e\",f\"{qtl_type}_mod_pos\",\"score\",\"strand\"]]\n",
    "        bed_df.to_csv(f'{qtl_type}_tmp.bed', sep='\\t', index=False, header=False)\n",
    "        call(f\"bedtools intersect -a {qtl_type}_tmp.bed -b {gene_bed_file} -wa -wb -s > {qtl_type}_gene.bed\", shell=True)\n",
    "        gene_df = pd.read_csv(f'{qtl_type}_gene.bed', sep='\\t', header=None, usecols=[0,3,5,9])\n",
    "        gene_df.columns = [\"chrom\",\"mod_pos\",\"strand\",\"gene_id\"]\n",
    "        gene_df[\"name\"] = gene_df['chrom']+\"_\"+gene_df['mod_pos'].astype(str)\n",
    "        gene_df['chrom'] = gene_df['chrom'].str.replace('chr', '', regex=False)\n",
    "        gene_df[\"LD_pos\"] = 0\n",
    "        res_df = gene_df[[\"chrom\",\"name\",\"LD_pos\",\"mod_pos\",\"gene_id\",\"strand\"]]\n",
    "        call(f\"rm {qtl_type}_tmp.bed\", shell=True)\n",
    "        flist_df = res_df.copy()\n",
    "        flist_df['PathOfEsd'] = f'{work_dir}/{qtl_type}_'+ flist_df['mod_pos'].astype(str) +'.esd'\n",
    "        flist_df.columns = [\"Chr\",\"ProbeID\",\"GeneticDistance\",\"ProbeBp\",\"Gene\",\"Orientation\",'PathOfEsd']\n",
    "        flist_df.to_csv(f'{qtl_type}.flist', sep='\\t', index=False)\n",
    "    elif qtl_type == \"stQTL\":\n",
    "        df['s'] = df[\"snp_pos\"].astype(int)-1\n",
    "        df['e'] = df[\"snp_pos\"].astype(int)+1\n",
    "        df['score'] = 0\n",
    "        df['rsID'] = df.apply(lambda row: f\"{row['chrom']}:{str(row['snp_pos'])}\" if row['rsID'] == \".\" else row['rsID'], axis=1)\n",
    "        bed_df = df[[\"chrom\",\"s\",\"e\",\"rsID\",\"score\",\"strand\"]]\n",
    "        bed_df.to_csv(f'{qtl_type}_tmp.bed', sep='\\t', index=False, header=False)\n",
    "        call(f\"bedtools intersect -a {qtl_type}_tmp.bed -b {gene_bed_file} -wa -wb -s > {qtl_type}_gene.bed\", shell=True)\n",
    "        gene_df = pd.read_csv(f'{qtl_type}_gene.bed', sep='\\t', header=None, usecols=[3,7,9])\n",
    "        gene_df.columns = [\"rsID\",\"gene_start\",\"gene_id\"]\n",
    "        qtl_data = pd.merge(qtl_data, gene_df, on=\"rsID\", how=\"left\")\n",
    "        # make esd\n",
    "        qtl_data['chrom'] = qtl_data['chrom'].str.replace('chr', '', regex=False)\n",
    "        dfs = qtl_data.groupby('gene_id')\n",
    "        for name, group in dfs:\n",
    "            res_df = group[[\"chrom\",\"rsID\",\"snp_pos\",\"A1\",\"A2\",\"EAF\",f\"{qtl_type}_Beta\",f\"{qtl_type}_se\",f\"{qtl_type}_pvalue\"]]\n",
    "            res_df.columns = [\"Chr\",\"SNP\",\"Bp\",\"A1\",\"A2\",\"Freq\",\"Beta\",\"se\",\"p\"]\n",
    "            res_df.to_csv(f'{qtl_type}_{name}.esd', index=False, sep=\"\\t\", header=False)\n",
    "        # make epi\n",
    "        qtl_data[\"LD_pos\"] = 0\n",
    "        res_df = qtl_data[[\"chrom\",\"rsID\",\"LD_pos\",\"gene_start\",\"gene_id\",\"strand\"]]\n",
    "        call(f\"rm {qtl_type}_tmp.bed\", shell=True)\n",
    "        # make flist\n",
    "        flist_df = res_df.copy()\n",
    "        flist_df['PathOfEsd'] = f'{work_dir}/{qtl_type}_'+ flist_df['gene_id'] +'.esd'\n",
    "        flist_df.columns = [\"Chr\",\"ProbeID\",\"GeneticDistance\",\"ProbeBp\",\"Gene\",\"Orientation\",'PathOfEsd']\n",
    "        flist_df.to_csv(f'{qtl_type}.flist', sep='\\t', index=False)\n",
    "    elif qtl_type == \"3aQTL\":\n",
    "        call(f\"bedtools intersect -a {apa_bed_file} -b {gene_bed_file} -wa -wb -s > apa_gene.bed\",shell=True)\n",
    "        apa_df = pd.read_csv(\"apa_gene.bed\",sep=\"\\t\",header=None,usecols=[0,1,3,5,13],names=[\"chrom\",\"apa_s\",\"apa_id\",\"strand\",\"gene_id\"])\n",
    "        apa_df = apa_df[apa_df['apa_id'].isin(df[f\"{qtl_type}_id\"])]\n",
    "        apa_df['chrom'] = apa_df['chrom'].str.replace('chr', '', regex=False)\n",
    "        apa_df['LD_pos'] = 0\n",
    "        res_df = apa_df[[\"chrom\",\"apa_id\",\"LD_pos\",\"apa_s\",\"gene_id\",\"strand\"]]\n",
    "        call(f\"rm apa_gene.bed\",shell=True)\n",
    "        flist_df = res_df.copy()\n",
    "        flist_df['PathOfEsd'] = f'{work_dir}/{qtl_type}_'+ flist_df['apa_id'] +'.esd'\n",
    "        flist_df.columns = [\"Chr\",\"ProbeID\",\"GeneticDistance\",\"ProbeBp\",\"Gene\",\"Orientation\",'PathOfEsd']\n",
    "        flist_df.to_csv(f'{qtl_type}.flist', sep='\\t', index=False)\n",
    "    elif qtl_type == \"irQTL\":\n",
    "        call(f\"bedtools intersect -a {isoform_bed_file} -b {gene_bed_file} -wa -wb -s > isoform_gene.bed\",shell=True)\n",
    "        isoform_df = pd.read_csv(\"isoform_gene.bed\",sep=\"\\t\",header=None,usecols=[0,1,3,5,15],names=[\"chrom\",\"isoform_s\",\"isoform_id\",\"strand\",\"gene_id\"])\n",
    "        isoform_df = isoform_df[isoform_df['isoform_id'].isin(df[f\"{qtl_type}_id\"])]\n",
    "        isoform_df['chrom'] = isoform_df['chrom'].str.replace('chr', '', regex=False)\n",
    "        isoform_df['LD_pos'] = 0\n",
    "        res_df = isoform_df[[\"chrom\",\"isoform_id\",\"LD_pos\",\"isoform_s\",\"gene_id\",\"strand\"]]\n",
    "        call(f\"rm isoform_gene.bed\",shell=True)\n",
    "        flist_df = res_df.copy()\n",
    "        flist_df['PathOfEsd'] = f'{work_dir}/{qtl_type}_'+ flist_df['isoform_id'] +'.esd'\n",
    "        flist_df.columns = [\"Chr\",\"ProbeID\",\"GeneticDistance\",\"ProbeBp\",\"Gene\",\"Orientation\",'PathOfEsd']\n",
    "        flist_df.to_csv(f'{qtl_type}.flist', sep='\\t', index=False)\n",
    "    else:\n",
    "        call(f\"bedtools intersect -a {pu_bed_file} -b {gene_bed_file} -wa -wb -s > promoter_gene.bed\",shell=True)\n",
    "        pu_df = pd.read_csv(\"promoter_gene.bed\",sep=\"\\t\",header=None,usecols=[0,1,3,5,9],names=[\"chrom\",\"pu_s\",\"pu_id\",\"strand\",\"gene_id\"])\n",
    "        pu_df = pu_df[pu_df['pu_id'].isin(df[f\"{qtl_type}_id\"])]\n",
    "        pu_df['chrom'] = pu_df['chrom'].str.replace('chr', '', regex=False)\n",
    "        pu_df['LD_pos'] = 0\n",
    "        res_df = pu_df[[\"chrom\",\"pu_id\",\"LD_pos\",\"pu_s\",\"gene_id\",\"strand\"]]\n",
    "        call(f\"rm promoter_gene.bed\",shell=True)\n",
    "        flist_df = res_df.copy()\n",
    "        flist_df['PathOfEsd'] = f'{work_dir}/{qtl_type}_'+ flist_df['pu_id'] +'.esd'\n",
    "        flist_df.columns = [\"Chr\",\"ProbeID\",\"GeneticDistance\",\"ProbeBp\",\"Gene\",\"Orientation\",'PathOfEsd']\n",
    "        flist_df.to_csv(f'{qtl_type}.flist', sep='\\t', index=False)\n",
    "    res_df.to_csv(f'{qtl_type}.epi', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL_all.flist\",sep=\"\\t\")\n",
    "df = df.head(10000)\n",
    "df.to_csv(\"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL.flist\",sep=\"\\t\",index=False)\n",
    "\n",
    "df = pd.read_csv(\"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL_all.epi\",sep=\"\\t\")\n",
    "df = df.head(10000)\n",
    "df.to_csv(\"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL.epi\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>ProbeID</th>\n",
       "      <th>GeneticDistance</th>\n",
       "      <th>ProbeBp</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>PathOfEsd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1229</th>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000718278.1-1-ENSG00000115539.15-10</td>\n",
       "      <td>0</td>\n",
       "      <td>101179454</td>\n",
       "      <td>ENSG00000238328.1</td>\n",
       "      <td>+</td>\n",
       "      <td>/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1230</th>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000718278.1-1-ENSG00000115539.15-10</td>\n",
       "      <td>0</td>\n",
       "      <td>101179454</td>\n",
       "      <td>ENSG00000115539.14_9</td>\n",
       "      <td>+</td>\n",
       "      <td>/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/M...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chr                                    ProbeID  GeneticDistance  \\\n",
       "1229    2  ENST00000718278.1-1-ENSG00000115539.15-10                0   \n",
       "1230    2  ENST00000718278.1-1-ENSG00000115539.15-10                0   \n",
       "\n",
       "        ProbeBp                  Gene Orientation  \\\n",
       "1229  101179454     ENSG00000238328.1           +   \n",
       "1230  101179454  ENSG00000115539.14_9           +   \n",
       "\n",
       "                                              PathOfEsd  \n",
       "1229  /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/M...  \n",
       "1230  /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/M...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/puqtl/promoter_final.bed\",sep=\"\\t\")\n",
    "# fourth_column = df.iloc[:, 3]\n",
    "# duplicate_rows = df[fourth_column.duplicated(keep=False)]\n",
    "# duplicate_rows = df[df.duplicated(subset=['PathOfEsd'], keep=False)]\n",
    "df = df[df['PathOfEsd']=='/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTLENST0000071827811ENSG000001155391510.esd']\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]\n",
       "Index: []"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('/mnt/hpc/home/xuxinran/DirectSeq/data/zhaolin_240206/240201-zhaolin-RNA-merge/v0.8.1/irqtl/nano_merge.annotated_transcripts.bed',sep=\"\\t\",header=None)\n",
    "fourth_column = df.iloc[:, 3]\n",
    "df = df[fourth_column=='ENST00000718278.1-1,ENSG00000115539.15-10']\n",
    "dfENST00000718278.1_1_ENSG00000115539.15_10\n",
    "\n",
    "ENST00000718278.1_1_ENSG00000115539.15_10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict7 = {\n",
    "    \"m6AQTL\": \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL\",\n",
    "    \"pseUQTL\": \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL\",\n",
    "    \"m5CQTL\": \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL\",\n",
    "    \"inosineQTL\": \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL\",\n",
    "    \"puQTL\": \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL\",\n",
    "    \"3aQTL\": \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL\",\n",
    "    \"irQTL\": \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_3aQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL_irQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m6AQTL/m6AQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_m6AQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/pseUQTL/pseUQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_pseUQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/m5CQTL/m5CQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_m5CQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/inosineQTL/inosineQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_inosineQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/puQTL/puQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_puQTL --thread-num 10 \n",
      "/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL/irQTL --beqtl-summary /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/3aQTL/3aQTL --out /mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/irQTL_3aQTL --thread-num 10 \n"
     ]
    }
   ],
   "source": [
    "def cross_combinations(list1, list2):\n",
    "    result = []\n",
    "    for x, y in itertools.product(list1, list2):\n",
    "        if x != y:  # 添加判断条件，跳过 x 和 y 相同的组合\n",
    "            result.append(f\"{x}-{y}\")\n",
    "    for y, x in itertools.product(list2, list1):\n",
    "        if y != x: # 添加判断条件，跳过 y 和 x 相同的组合\n",
    "            result.append(f\"{y}-{x}\")\n",
    "    return result\n",
    "\n",
    "qtl_types = list(dict7.keys())\n",
    "result = cross_combinations(qtl_types, qtl_types)\n",
    "work_dir = '/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR'\n",
    "for i in result:\n",
    "    qtl1 = i.split(\"-\")[0]\n",
    "    qtl2 = i.split(\"-\")[1]\n",
    "    qtl1_file = dict7[qtl1]\n",
    "    qtl2_file = dict7[qtl2]\n",
    "    print(f'/mnt/hpc/home/xuxinran/GWAS/smr-1.3.1-linux-x86_64/smr --bfile /mnt/hpc/home/xuxinran/huvec_genotype/huvec_imputed --beqtl-summary {qtl1_file} --beqtl-summary {qtl2_file} --out {work_dir}/{qtl1}_{qtl2} --thread-num 10 ')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 结果解读 处理SMR的结果\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def merge_smr_files(folder_path):\n",
    "    all_data = []\n",
    "    for filename in os.listdir(folder_path):\n",
    "        if filename.endswith(\".smr\"):\n",
    "            file_path = os.path.join(folder_path, filename)\n",
    "            qtl1 = filename.split(\"_\")[0]\n",
    "            qtl2 = filename.split(\"_\")[1].replace(\".smr\",\"\")\n",
    "            try:\n",
    "                df = pd.read_csv(file_path, sep='\\t',usecols=['Expo_ID','Expo_Chr','Expo_Gene','Outco_ID','Outco_Chr','Outco_Gene','topSNP','b_SMR', 'se_SMR', 'p_SMR'])\n",
    "                n = len(df)\n",
    "                if n == 0:\n",
    "                    continue\n",
    "                df = df[df['p_SMR']<(0.05/n)]\n",
    "                df['Expo_type'] = qtl1\n",
    "                df['Outco_type'] = qtl2\n",
    "                if df.empty:\n",
    "                    continue\n",
    "                all_data.append(df)\n",
    "            except Exception as e:\n",
    "                print(f\"读取文件 {filename} 时出错: {e}\")\n",
    "\n",
    "    if all_data:\n",
    "        merged_df = pd.concat(all_data, ignore_index=True)\n",
    "        return merged_df\n",
    "\n",
    "# 示例用法\n",
    "folder_path = \"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/\"  # 替换为您的文件夹路径\n",
    "\n",
    "\n",
    "a = merge_smr_files(folder_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Expo_ID</th>\n",
       "      <th>Expo_Chr</th>\n",
       "      <th>Expo_Gene</th>\n",
       "      <th>Outco_ID</th>\n",
       "      <th>Outco_Chr</th>\n",
       "      <th>Outco_Gene</th>\n",
       "      <th>topSNP</th>\n",
       "      <th>b_SMR</th>\n",
       "      <th>se_SMR</th>\n",
       "      <th>p_SMR</th>\n",
       "      <th>Expo_type</th>\n",
       "      <th>Outco_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TMEM43-1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000170876.9_8;ENSG00000268279.4_14;ENSG0...</td>\n",
       "      <td>chr3-14184315</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000170876.9_8;ENSG00000268279.4_14;ENSG0...</td>\n",
       "      <td>rs116911972</td>\n",
       "      <td>-0.727548</td>\n",
       "      <td>0.172006</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>m6AQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TMEM43-3-TMEM43-2</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000170876.9_8;ENSG00000268279.4_14;ENSG0...</td>\n",
       "      <td>chr3-14184315</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000170876.9_8;ENSG00000268279.4_14;ENSG0...</td>\n",
       "      <td>rs116911972</td>\n",
       "      <td>0.727548</td>\n",
       "      <td>0.172006</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>m6AQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PXN-2</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>chr12-120650014</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>rs4767884</td>\n",
       "      <td>0.882719</td>\n",
       "      <td>0.218328</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>m6AQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PXN-1</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>chr12-120650014</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>rs4767884</td>\n",
       "      <td>-0.881969</td>\n",
       "      <td>0.217827</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>m6AQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BCL2L2-1-BCL2L2-2-BCL2L2-3</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...</td>\n",
       "      <td>chr14-23790840</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...</td>\n",
       "      <td>rs2295126</td>\n",
       "      <td>1.379020</td>\n",
       "      <td>0.377352</td>\n",
       "      <td>0.000258</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>m6AQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048</th>\n",
       "      <td>PXN-1</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>chr12-120651706</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>rs4767884</td>\n",
       "      <td>-0.617075</td>\n",
       "      <td>0.159256</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>pseUQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1049</th>\n",
       "      <td>PXN-2</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>chr12-120652777</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>rs4767884</td>\n",
       "      <td>0.618000</td>\n",
       "      <td>0.159211</td>\n",
       "      <td>0.000104</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>pseUQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1050</th>\n",
       "      <td>PXN-1</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>chr12-120652777</td>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000089159.17_19</td>\n",
       "      <td>rs4767884</td>\n",
       "      <td>-0.617476</td>\n",
       "      <td>0.158864</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>pseUQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1051</th>\n",
       "      <td>BCL2L2-1-BCL2L2-2-BCL2L2-3</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...</td>\n",
       "      <td>chr14-23791019</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...</td>\n",
       "      <td>rs2295126</td>\n",
       "      <td>1.263460</td>\n",
       "      <td>0.310626</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>pseUQTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1052</th>\n",
       "      <td>PABPN1-1-PABPN1-4-PABPN1-2-PABPN1-7-PABPN1-6-P...</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...</td>\n",
       "      <td>chr14-23791019</td>\n",
       "      <td>14</td>\n",
       "      <td>ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...</td>\n",
       "      <td>rs2295126</td>\n",
       "      <td>-1.263840</td>\n",
       "      <td>0.310986</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>puQTL</td>\n",
       "      <td>pseUQTL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1053 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Expo_ID  Expo_Chr  \\\n",
       "0                                              TMEM43-1         3   \n",
       "1                                     TMEM43-3-TMEM43-2         3   \n",
       "2                                                 PXN-2        12   \n",
       "3                                                 PXN-1        12   \n",
       "4                            BCL2L2-1-BCL2L2-2-BCL2L2-3        14   \n",
       "...                                                 ...       ...   \n",
       "1048                                              PXN-1        12   \n",
       "1049                                              PXN-2        12   \n",
       "1050                                              PXN-1        12   \n",
       "1051                         BCL2L2-1-BCL2L2-2-BCL2L2-3        14   \n",
       "1052  PABPN1-1-PABPN1-4-PABPN1-2-PABPN1-7-PABPN1-6-P...        14   \n",
       "\n",
       "                                              Expo_Gene         Outco_ID  \\\n",
       "0     ENSG00000170876.9_8;ENSG00000268279.4_14;ENSG0...    chr3-14184315   \n",
       "1     ENSG00000170876.9_8;ENSG00000268279.4_14;ENSG0...    chr3-14184315   \n",
       "2                                 ENSG00000089159.17_19  chr12-120650014   \n",
       "3                                 ENSG00000089159.17_19  chr12-120650014   \n",
       "4     ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...   chr14-23790840   \n",
       "...                                                 ...              ...   \n",
       "1048                              ENSG00000089159.17_19  chr12-120651706   \n",
       "1049                              ENSG00000089159.17_19  chr12-120652777   \n",
       "1050                              ENSG00000089159.17_19  chr12-120652777   \n",
       "1051  ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...   chr14-23791019   \n",
       "1052  ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...   chr14-23791019   \n",
       "\n",
       "      Outco_Chr                                         Outco_Gene  \\\n",
       "0             3  ENSG00000170876.9_8;ENSG00000268279.4_14;ENSG0...   \n",
       "1             3  ENSG00000170876.9_8;ENSG00000268279.4_14;ENSG0...   \n",
       "2            12                              ENSG00000089159.17_19   \n",
       "3            12                              ENSG00000089159.17_19   \n",
       "4            14  ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...   \n",
       "...         ...                                                ...   \n",
       "1048         12                              ENSG00000089159.17_19   \n",
       "1049         12                              ENSG00000089159.17_19   \n",
       "1050         12                              ENSG00000089159.17_19   \n",
       "1051         14  ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...   \n",
       "1052         14  ENSG00000129473.11_16;ENSG00000258643.6_14;ENS...   \n",
       "\n",
       "           topSNP     b_SMR    se_SMR     p_SMR Expo_type Outco_type  \n",
       "0     rs116911972 -0.727548  0.172006  0.000023     puQTL     m6AQTL  \n",
       "1     rs116911972  0.727548  0.172006  0.000023     puQTL     m6AQTL  \n",
       "2       rs4767884  0.882719  0.218328  0.000053     puQTL     m6AQTL  \n",
       "3       rs4767884 -0.881969  0.217827  0.000051     puQTL     m6AQTL  \n",
       "4       rs2295126  1.379020  0.377352  0.000258     puQTL     m6AQTL  \n",
       "...           ...       ...       ...       ...       ...        ...  \n",
       "1048    rs4767884 -0.617075  0.159256  0.000107     puQTL    pseUQTL  \n",
       "1049    rs4767884  0.618000  0.159211  0.000104     puQTL    pseUQTL  \n",
       "1050    rs4767884 -0.617476  0.158864  0.000102     puQTL    pseUQTL  \n",
       "1051    rs2295126  1.263460  0.310626  0.000048     puQTL    pseUQTL  \n",
       "1052    rs2295126 -1.263840  0.310986  0.000048     puQTL    pseUQTL  \n",
       "\n",
       "[1053 rows x 12 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "a.to_csv(\"/mnt/hpc/home/xuxinran/DirectSeq/8_downsteam/MR/smr_result.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://yanglab.westlake.edu.cn/software/smr/#SMRlocusplot19\n",
    "画图"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "methy1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
